### ISSUE TITLE
[Enhancement]: DNN behind Cloudflare - HTTP_X_FORWARDED_FOR

### ISSUE THREAD

**Issue by fablaser** [dnnsoftware/Dnn.Platform#6145]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description of problem

Hello, I have some websites behind Cloudflare's WAF. I currently use Request.ServerVariables("HTTP_X_FORWARDED_FOR") for my own applicatoin log. So I can log real IP address of clients, not just Cloudflare ip addresses. It would be useful to have an option that allows DNN to log IP using this variable.

### Description of solution

implement Request.ServerVariables("HTTP_X_FORWARDED_FOR")

### Do you be plan to contribute code for this enhancement?

- [ ] Yes

### Would you be interested in sponsoring this enhancement?

- [X] Yes

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by jeremy-farrance** [comment]
Related? to #5073 ?

This sounds like an enhancement that a lot of people would appreciate. Well, at least me!

**Comment by valadas** [comment]
@fablaser can you try setting it here and see if that works ?

![image](https://github.com/user-attachments/assets/88bd2400-9847-4d99-9fc9-a7a8385dffea)

**Comment by fablaser** [comment]
> @fablaser can you try setting it here and see if that works ?
> 
> ![image](https://private-user-images.githubusercontent.com/6371568/372569339-88bd2400-9847-4d99-9fc9-a7a8385dffea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc4NTM1MjcsIm5iZiI6MTcyNzg1MzIyNywicGF0aCI6Ii82MzcxNTY4LzM3MjU2OTMzOS04OGJkMjQwMC05ODQ3LTRkOTktOWZjOS1hN2E4Mzg1ZGZmZWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMDJUMDcxMzQ3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzY1ZWYzNTZiZGFmNjAzYTdhZjQ0YWZjMGYxZGNiMzc3N2ExOWE1YjFlMzYxMDllMzg0NDdhMzUwNTE0YzcxYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.GTK-4UVaM6oK2beHaZQ1F8YjC0QZtyvnojfq3dmHp8Y)

Hello, 
I tried to follow the configuration, but I does not work... After changed it I tried to clear cache and recycle app pool. Anyway any operation (such as failed login) will be logged with Cloudflare ip address...

**Comment by fablaser** [comment]
> Related? to #5073 ?
> 
> This sounds like an enhancement that a lot of people would appreciate. Well, at least me!

Yes, it is related to it! I did not noticed this previous request

**Comment by valadas** [comment]
Just reading some code, it looks like one could add an entry in the HostSettings table to specify that header

https://github.com/dnnsoftware/Dnn.Platform/blob/9f4833f5dcba24e0dbd194338c2a50a03a43da83/DNN%20Platform/Library/Services/UserRequest/UserRequestIPAddressController.cs#L27C1-L28C1

**Comment by fablaser** [comment]
Hello
thank you for your support.

I tried to add a new record in HostSettings table with these values:

SettingName: UserRequestIPHeader
SettingValue: X-Forwarded-For

Then I cleared cache and recycled app pool, but the result is always the same: every failed login, in logs, has the cloudflare ip address...

**Comment by fablaser** [comment]
> Just reading some code, it looks like one could add an entry in the HostSettings table to specify that header
> 
> https://github.com/dnnsoftware/Dnn.Platform/blob/9f4833f5dcba24e0dbd194338c2a50a03a43da83/DNN%20Platform/Library/Services/UserRequest/UserRequestIPAddressController.cs#L27C1-L28C1

Hello, 
I simply needed to write "X-Forwarded-For" in lower case!
Thanks!!

**Comment by jeremy-farrance** [comment]
> 
> Hello, 
> I simply needed to write "X-Forwarded-For" in lower case!

Wait, what? Are you saying, 

- in code you need it lower cased... Request.ServerVariables() or 

- the Value it the SQL table HostSettings

needed to be lowercase?

**Comment by valadas** [comment]
That is surprising, they document it with capitals https://developers.cloudflare.com/fundamentals/reference/http-request-headers/#x-forwarded-for

**Comment by fablaser** [comment]
> Wait, what? Are you saying,
>     * in code you need it lower cased... Request.ServerVariables() or
>     * the Value it the SQL table HostSettings
> 
> 
> needed to be lowercase?

In SQL HostSettings table. When DNN launces this code:

```
 var userRequestIPHeader = HostController.Instance.GetString("UserRequestIPHeader", "X-Forwarded-For");
 var userIPAddress = string.Empty;
 if (request.Headers.AllKeys.Contains(userRequestIPHeader))

```
AllKeys.Contains(userRequestIPHeader)) does not matches "X-Forwarded-For", it is seems to be case sensitive.

**Comment by fablaser** [comment]
> That is surprising, they document it with capitals https://developers.cloudflare.com/fundamentals/reference/http-request-headers/#x-forwarded-for

Perhaps it should be useful to change the code using Contains(userRequestIPHeader, IEqualityComparer) so that it would be case insensitive...

**Comment by valadas** [comment]
Oh, I see, so you went from `X-FORWARDED-FOR` to `X-Forwarded-For` right ?

**Comment by ronnydodd** [comment]
I wont ask again, please take me off your mailing list, NOW!


Frank DeSocio
Bpaa Executive Director
817-385-8428
316-648-2479   cell

**Comment by fablaser** [comment]
> Oh, I see, so you went from `X-FORWARDED-FOR` to `X-Forwarded-For` right ?

In my own code I used X-FORWARDED-FOR, but it does not matter if I use request.Headers['X-FORWARDED-FOR'], request.Headers['X-Forwarded-For'] or request.Headers['x-forwarded-for'], they will work fine. 

The issue has been caused by "request.Headers.AllKeys.Contains(userRequestIPHeader)" in DNN code you mentioned yesterday. "Contains" method is case sensitive. In my header's collection I see "x-forwarded-for" (lower), so I need to use this string in hostsettings table.

**Comment by MarietteNL** [comment]
INSERT INTO [dbo].[HostSettings]
           ([SettingName],[SettingValue],[SettingIsSecure],[CreatedByUserID],[CreatedOnDate],[LastModifiedByUserID],[LastModifiedOnDate])
     VALUES ('UserRequestIPHeader','x-forwarded-for',0,-1,GETDATE(),-1,GETDATE())
GO
